ee/middleware/upstreamoauth/provider.go

```diff
@@ -0,0 +1,318 @@
+package upstreamoauth
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/TykTechnologies/tyk/ctx"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	oauth2clientcredentials "golang.org/x/oauth2/clientcredentials"
+
+	"github.com/TykTechnologies/tyk/apidef"
+	"github.com/TykTechnologies/tyk/internal/crypto"
+	"github.com/TykTechnologies/tyk/internal/event"
+	"github.com/TykTechnologies/tyk/internal/model"
+	"github.com/TykTechnologies/tyk/storage"
+)
+
+// Provider implements upstream auth provider.
+type Provider struct {
+	// Logger is the logger to be used.
+	Logger *logrus.Entry
+	// HeaderName is the header name to be used to fill upstream auth with.
+	HeaderName string
+	// AuthValue is the value of auth header.
+	AuthValue string
+}
+
+// Fill sets the request's HeaderName with AuthValue
+func (u Provider) Fill(r *http.Request) {
+	if r.Header.Get(u.HeaderName) != "" {
+		u.Logger.WithFields(logrus.Fields{
+			"header": u.HeaderName,
+		}).Info("Authorization header conflict detected: Client header overwritten by Gateway upstream authentication header.")
+	}
+	r.Header.Set(u.HeaderName, u.AuthValue)
+}
+
+type OAuthHeaderProvider interface {
```

Comment:
This isn't flexible, if anything the implementations for the interface should be:

1. inside a ee/middleware/upstreamoauth/internal scope,
2. be fully exported symbols for each member of the interface

Them not being exported means only local mocking or implementations are possible. To decouple scopes, we need interfacing symbols exported.

Reply:
Agreed to tackle this at a later date.